In [3]:
import numpy as np
import matplotlib.pyplot as plt

def f(x):
    return 100 - 2 * x**2 + 3 * x

def maximizer(a, b, N, f):
    x = np.linspace(a, b, N)
    y = f(x)
    max_index = np.argmax(y)
    maximum = y[max_index]
    maximizer_value = x[max_index]
    return {'maximum': maximum, 'maximizer': maximizer_value}

a = -1
b = 1
print(maximizer(a,b,3,f))
print(maximizer(a,b,10,f))
print(maximizer(a,b,100,f))
print(maximizer(a,b,1000,f))
print(maximizer(a,b,5000,f))

f(.75)

{'maximum': 101.0, 'maximizer': 1.0}
{'maximum': 101.12345679012346, 'maximizer': 0.7777777777777777}
{'maximum': 101.12488521579431, 'maximizer': 0.7575757575757578}
{'maximum': 101.12499987474962, 'maximizer': 0.7497497497497498}
{'maximum': 101.124999994998, 'maximizer': 0.7499499899979996}


101.125

3. The value of f at x = 0.75 is not the same between the previous step and the actual value. While the actual maximum value is 101.125 at x = 0.75, the previous step returned a marginally smaller maximum value and a marginally smaller x-value. This is because the maximizer function is an estimation that is based on how many steps (N) there are. If N is small, the estimation will be less accurate as seen with the function returning the maximum as 101.0 and the maximizer as 0.777. If N is large, the estimation will be more accurate as seen with the function essentially returning the correct answer but not completely. As N approaches infinity (gets larger and larger), the maximum and maximizer approach the actual value.